pkg/client/client.go

```diff
@@ -1748,6 +1769,29 @@ func (c *Client) RegisterConsolePlugin(ctx context.Context, name string) error {
 	return nil
 }
 
+// VPACustomResourceDefinitionPresent checks if VerticalPodAutoscaler CRD is present in the cluster.
+func (c *Client) VPACustomResourceDefinitionPresent(ctx context.Context, lastKnownVPACustomResourceDefinitionPresent *bool) (*bool, error) {
+	_, err := c.ApiExtensionsInterface().ApiextensionsV1().CustomResourceDefinitions().Get(ctx, VerticalPodAutoscalerCRDMetadataName, metav1.GetOptions{})
+	if err != nil {
+		// VPA CRDs are absent.
+		if apierrors.IsNotFound(err) {
+			return ptr.To(false), nil
+		}
+
+		// See if we have an idea of the state of the CRD's presence before the transient error occurred.
+		// If we do, resort to that since we do not want this to cause unnecessary reconciles.
+		if lastKnownVPACustomResourceDefinitionPresent != nil {
+			return lastKnownVPACustomResourceDefinitionPresent, nil
+		}
+
+		// If we don't, throw.
+		return nil, fmt.Errorf("failed to get %s CRD: %w", VerticalPodAutoscalerCRDMetadataName, err)
+	}
+
+	// VPA CRDs are present.
+	return ptr.To(true), nil
+}
+
```

Comment:
I have two comments for this function:

1. Any reason not to return `(bool, err)`? Iiuc you make the bools to pointers here only to dereference them in [operator.go](https://github.com/openshift/cluster-monitoring-operator/pull/2078/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR767).
2. I'm not sure we should fall back to the last known return value here. Is adding/removing the VPA assets so expensive? I would argue that in an error scenario where resource retrieval doesn't work, our stack will likely have other problems.

Reply:
1. I remember the function signature accommodates for the three states of `lastKnownVPACustomResourceDefinitionPresent` that may be returned, i.e., `nil` (we don't remember the last state, as it was never set), `false` (we do, and they were absent), and `true` (we do, and they were present) (this was discussed in a call with @simonpasquier way back).
2. So the reason we had to check for VPA CRD presence before enabling them is due to the fact that KSM will complain (PTAL at Simon's comment here: https://github.com/openshift/cluster-monitoring-operator/pull/2078#issuecomment-1825652771) about any enabled GVKs that are absent from the cluster.

Please correct me if I misunderstood either of these.

Reply:
Sure I get the need to check and that this check for nil as a signal that there is no cache value. Not sure I understand the need for caching this value though. If that has been discussed before, maybe we can add the reasoning as a comment for future readers?

Reply:
I was able to dig through the reviews and find a textual version of our discussion here: https://github.com/openshift/cluster-monitoring-operator/pull/2078#discussion_r1410364467.

Right, Kube API failing resource retrieval operations will most likely have more severe repercussions stack-wide than this. But I think, to the best of my understanding (and I maybe going out on a limb here), Simon meant the _transient_ error to occur within the span of a couple ticks (whilst this retrieval is happening) so it may not affect other concurrent tasks that are not retrieving resources during those same ticks.

LMK if you still think we should drop this and I'll patch it up.

PS. I've tried to sum things up in a comment here: https://github.com/openshift/cluster-monitoring-operator/pull/2078/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR194-R212.

Reply:
:+1: thanks for the context!